feat(Scroll): Add syncswap AMM support

zinkkrysty opened the pull request:
Adds support for decoding syncswap transactions (swap, liquidity add/remove). Syncswap is a uniswap V2-like protocol, so a common evm module was added for decoding uniswap V2-like protocols. The intention is to use this for uniswap v2 and sushiswap, however that would have made the scope of this PR too big.

https://syncswap.gitbook.io/api-documentation

OjusWiZard commented:
also please add a changelog entry and rebase the PR on the latest develop branch

zinkkrysty commented on rotkehlchen/chain/evm/decoding/uniswap_v2_like/decoder.py:
```diff
@@ -0,0 +1,140 @@
+log = RotkehlchenLogsAdapter(logger)
+
+
+class UniswapV2LikeDecoder(DecoderInterface):
```
it's not used, but it would make sense to use it for `chain/ethereum/modules/uniswap/v2/decoder.py` and `rotkehlchen/chain/ethereum/modules/sushiswap/decoder.py`

Though I am unsure if doing the work in the `post_decoding_rules` as I am doing here is better or worse than using normal decoding rules. It's just that for syncswap the order of the events (I think for the deposit/withdraw) was reversed so I needed to access the decoded transfer events at the beginning.

zinkkrysty replied:
I know that you can't always abstract similar logic and trying to fit everything in one box is not always feasible, and can lead to performance problems. However I also think that a good principle to strive for is to abstract as much as possible, so that the common logic for decoding swaps would work in principle for any AMM, so that we can add add many AMMs as possible using as little code as possible. In some cases (such as for any AMM based on Uniswap V2 which is very popular), we might want to resort to the common denominator (in this case using post-decoding) if that would work for most AMM contracts.

LefterisJP replied:
post decoding rules is the worst type of rule for performance. I would not use it in an abstraction.

zinkkrysty commented on rotkehlchen/chain/evm/decoding/uniswap_v2_like/common.py:
```diff
@@ -225,37 +283,48 @@ def decode_uniswap_like_deposit_and_withdrawals(
-        factory_address=factory_address,
-        init_code_hash=init_code_hash,
+    pool_address = compute_pool_address(
+        token0,
+        token1,
+        factory_address,
+        init_code_hash,
```
no, unfortunately it does not work
![image](https://github.com/rotki/rotki/assets/570771/d03c5162-55f6-4b4b-843d-cb2e5b0d8e09)

zinkkrysty commented:
@OjusWiZard sorry for the long delay 😬. I rebased now (also the test-caching branch) and addressed most comments. I am not in a hurry to get this out. If you think the design of `UniswapV2LikeDecoder` and the common methods could be improved, I can take a stab at that - but will need some directions.

OjusWiZard reviewed (commented):
heyy @zinkkrysty, some very small requests

zinkkrysty commented:
@OjusWiZard thanks again for the review. I addressed all the suggestions

LefterisJP commented on rotkehlchen/chain/evm/decoding/uniswap_v2_like/decoder.py:
```diff
@@ -0,0 +1,140 @@
+
+
+class UniswapV2LikeDecoder(DecoderInterface):
+    weth_asset: EvmToken | None = None
```
why make this a part of the class and initialize to None? This should be set by the constructor and should always have a value. That's not how inheritance works.

LefterisJP commented on rotkehlchen/chain/scroll/modules/syncswap/decoder.py:
```diff
@@ -0,0 +1,60 @@
+SYNCSWAP_ROUTER_ADDRESS: Final = string_to_evm_address('0x80e38291e06339d10AAB483C65695D004dBD5C69')  # noqa: E501
+SYNCSWAP_VAULT_ADDRESS: Final = string_to_evm_address('0x7160570BB153Edd0Ea1775EC2b2Ac9b65F1aB61B')
+CPT_SYNCSWAP: Final = 'syncswap'
+SYNCSWAP_LABEL: Final = 'Syncswap'
```
Any variable that's only used once should not exist. Please for this and all other files just put them as literal constants where they are used.

Also anythign that's supposed to be a constant should be in the module's `constants/py`

LefterisJP commented on rotkehlchen/chain/scroll/modules/syncswap/decoder.py:
```diff
@@ -0,0 +1,60 @@
+SYNCSWAP_ICON: Final = 'syncswap.svg'
+SYNCSWAP_CLASSIC_POOL_FACTORY_ADDRESS: Final = string_to_evm_address('0x37BAc764494c8db4e54BDE72f6965beA9fa0AC2d')  # noqa: E501
+SYNCSWAP_STABLE_POOL_FACTORY_ADDRESS: Final = string_to_evm_address('0xE4CF807E351b56720B17A59094179e7Ed9dD3727')  # noqa: E501
+# The init code hash is looked up using the RPC method debug_traceTransaction() from the pool initialization transaction  # noqa: E501
+# The pool_bytecode is the input of the CREATE2 call, then we pass it to Web3.keccak(hexstring_to_bytes(pool_bytecode))  # noqa: E501
```
Don't put noqa: E501 in comments unless it's absolutely needed. Just break into more lines.

```suggestion
# The init code hash is looked up using the RPC method debug_traceTransaction() from the
#  pool initialization transaction. The pool_bytecode is the input of the CREATE2 call, then
# we pass it to Web3.keccak(hexstring_to_bytes(pool_bytecode))
```

LefterisJP commented on rotkehlchen/chain/scroll/modules/syncswap/decoder.py:
```diff
@@ -0,0 +1,60 @@
+                SYNCSWAP_STABLE_POOL_FACTORY_ADDRESS: SYNCSWAP_STABLE_INIT_CODE_HASH,
+            },
+        )
+        self.weth_asset = A_WETH_SCROLL.resolve_to_evm_token()
```
If you needed to use inheritance here this shoudl simply have beeen part of super().__init__().

LefterisJP commented on rotkehlchen/chain/evm/decoding/uniswap_v2_like/decoder.py:
```diff
@@ -0,0 +1,140 @@
+        else:
+            return
+
+        if self.weth_asset is None:
```
this if does not need to exist. It's always true. If you use inheritance properly.

LefterisJP commented on rotkehlchen/tests/unit/decoders/test_syncswap.py:
```diff
@@ -0,0 +1,336 @@
+            location_label=scroll_accounts[0],
+            notes=f'Burned {fee} ETH for gas',
+            counterparty='gas',
+        ),
+        EvmEvent(
```
Here and in the rest of the file.

```suggestion
        ), EvmEvent(
```

LefterisJP commented on rotkehlchen/tests/unit/decoders/test_syncswap.py:
```diff
@@ -0,0 +1,336 @@
+        database=database,
+        tx_hash=evmhash,
+    )
+    timestamp = TimestampMS(1713530626000)
+    fee = FVal('0.00024694228625118')
+    amount_received_usdc = FVal('36.508481')
+    amount_sent_eth = FVal('0.01180498459310175')
```
Here and in the rest of the file. Please keep style consistent.


```suggestion
    timestamp, fee, amount_received_usdc, amount_sent_eth = TimestampMS(.....
```

LefterisJP reviewed (commented):
Lots of comments. Also needs rebase and adjusting changelog. In general make small PRs, succint and keeping consistency with the rest of the codebase.

Abstractions can come later. Otherwise it's impossible to review and extremely time consuming to merge.
